crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_map.rs

```diff
@@ -67,169 +64,174 @@ impl Violation for UnnecessaryMap {
+                | Expr::List(ast::ExprList { elts, .. })) = body
+                else {
+                    return;
+                };
 
-            if !keywords.is_empty() {
-                return;
+                if elts.len() != 2 {
+                    return;
+                }
             }
 
-            let Some(argument) = helpers::first_argument_with_matching_function("map", func, args)
-            else {
-                return;
-            };
+            (lambda, parameters, body, iterables)
+        }
+    };
 
-            let Expr::Lambda(ast::ExprLambda {
-                parameters, body, ..
-            }) = argument
-            else {
-                return;
-            };
+    // For example, (x+1 for x in (c:=a)) is invalid syntax
+    // so we can't suggest it.
+    if iterables
+        .iter()
+        .any(|iterable| any_over_expr(iterable, &|expr| expr.is_named_expr()))
+    {
+        return;
+    }
 
-            let (Expr::Tuple(ast::ExprTuple { elts, .. }) | Expr::List(ast::ExprList { elts, .. })) =
-                body.as_ref()
-            else {
-                return;
-            };
+    if iterables.iter().any(Expr::is_starred_expr) {
+        return;
+    }
 
-            if elts.len() != 2 {
-                return;
-            }
+    if checker.settings.preview.is_disabled() && iterables.len() != 1 {
+        return;
+    }
 
-            if !lambda_has_expected_arity(parameters.as_deref(), body) {
-                return;
-            }
-        }
-    };
+    if !lambda_has_expected_arity(parameters, body, iterables) {
+        return;
+    }
 
-    let mut diagnostic = Diagnostic::new(UnnecessaryMap { object_type }, expr.range());
+    let mut diagnostic = Diagnostic::new(UnnecessaryMap { object_type }, call.range);
     diagnostic.try_set_fix(|| {
-        fixes::fix_unnecessary_map(
-            expr,
-            parent,
-            object_type,
-            checker.locator(),
-            checker.stylist(),
-        )
-        .map(Fix::unsafe_edit)
+        let edit = if checker.settings.preview.is_enabled() {
+            fixes::fix_unnecessary_map_preview(
+                call.range,
+                parent,
+                object_type,
+                lambda,
+                parameters,
+                body,
+                iterables,
+                checker,
+            )
+        } else {
+            fixes::fix_unnecessary_map(
+                call.range,
+                parent,
+                object_type,
+                checker.locator(),
+                checker.stylist(),
+            )
+        };
+
+        edit.map(Fix::unsafe_edit)
     });
     checker.diagnostics.push(diagnostic);
 }
 
+fn is_list_set_or_dict(func: &Expr, semantic: &SemanticModel) -> bool {
+    semantic.match_builtin_expr(func, "list")
+        || semantic.match_builtin_expr(func, "set")
+        || semantic.match_builtin_expr(func, "dict")
```

Comment:
This seems inefficient, because we have to resolve the qualified name multiple times. Something like this might be cheaper:

```suggestion
    semantic
        .resolve_qualified_name(func)
        .is_some_and(|qualified_name| {
            matches!(
                qualified_name.segments(),
                ["" | "buitins", "list" | "set" | "dict"]
            )
        })
```